Refactoring

kevin-brown opened the pull request:
> Originally at https://github.com/kevin-brown/mousetrap/pull/1

This is the pull request that will be used for code review, tracking progress, and eventually generating the patch set for this refactoring.

StoneyJackson commented:
Use this pull-request to review code, ask questions, get an overview of what is going on, etc. When you commit code, write useful commit comments, so folks can follow along (that goes for me too!).

## Goal

We are trying to improve the design and implementation through refactoring. As Heidi recently pointed out, this is turning into a rewrite.

## Design

The diagrams below gives an overview of what we are trying to do. However, as we work, we expect the design to evolve. So use these as a starting point, not as law, and not as definitive documentation of the code. Ultimately code is king.
- [Use cases and class descriptions](https://wiki.gnome.org/Projects/MouseTrap/UseCases)
- [Collaboration diagram for Identify Tracking Points](https://plus.google.com/+StoneyJackson1/posts/b2Cy5AM1ng9)
- [Collaboration diagram for Track Changes](https://plus.google.com/+StoneyJackson1/posts/4djiQkBymH3)
- [Diagram of previous file hierarchy](https://plus.google.com/+StoneyJackson1/posts/fPx8zovUgfR)
- [Diagram of new file hierarchy](https://plus.google.com/+StoneyJackson1/posts/PQ6KV24W7C5)

StoneyJackson commented:
2014-05-16 (HE):  Refactoring is becoming almost development from scratch. Stoney is updating a revised Camera class. Please use this as an example for naming and form. The current approach is to simply try to get steps 1, 2, and 6 from the Identify Tracking Points use case to work. This is getting an image from the camera and displaying the image.

kevin-brown commented:
I started back working on MouseTrap today, and it's interesting to see the path it's heading in.  Given that I'm using a debian environment, I did have to fix an issue with the `/bin/env` shebangs, which was committed @ https://github.com/GNOME-MouseTrap/mousetrap/commit/801e8fd5b76d53813ffac97f900c80959c061224.

The other issue I hit was with PyLint because I use a symlink in my directory path.  For some reason, PyLint cannot correctly trace modules if the path that is provided contains a symlink.  This usually wouldn't be an issue, except the snippet we are using to get the `MOUSETRAP_HOME` does not account for this.  There is an answer on StackOverflow with the original snippet that [points out this issue and gives a solution](http://stackoverflow.com/a/246128), so that might need to be considered.  It is easier (and less confusion) to do it in Python, which may also be an option considering Python is required for MouseTrap.

I can resolve the symlink and make it work, but I wouldn't expect developers to do it.

StoneyJackson commented:
OK... I finally posted what I worked on while camping Memorial Day Weekend. I'll defer to Kevin on how to handle pylint. Kevin, I'm not sure I know what's best and why. So feel free to make more adjustments and/or give us guidelines on what we should be doing.

kevin-brown commented on src/mousetrap/vision.py:
```diff
@@ -0,0 +1,37 @@
+        self.device = self._new_capture_device(device_index)
+        self._set_dimensions(width, height)
+
+    @staticmethod
```
Is there any reason why this method is marked as static?  It's unusual for private methods to be static and I can't see why it is being done in this case.

StoneyJackson replied:
Short answer: after refactoring and cleaning, I noticed that this method doesn't use the state of an instance. That sounds like a static method to me. So I made it static. It's private because only code in the class uses it (for now).

Longer answer: if something naturally looks static (as described above) I prefer to make it so. Static methods are often prime candidates for reuse. And with less dependencies, they are in a better position to be refactored into a module function or another "library" class.

Deeper issues: if one finds many private static methods in a class, it could be a sign that we are not programming in an OO fashion. Hiding the issue by making all methods non-static would not change that fact, it would just make it harder to detect. In my mind, I didn't make it static, it was already static. I just labeled it as such.

Very deep issues: these are private :)

kevin-brown commented on src/mousetrap/vision.py:
```diff
@@ -0,0 +1,37 @@
+        return capture
+
+    def _set_dimensions(self, width, height):
+        self.device.set(cv.CV_CAP_PROP_FRAME_WIDTH, width)
```
`cv2.cv` is not available in OpenCV 3 (it is technically the old `cv` library).  Sadly these parameters are no longer defined as "constants", so we may need to redefine them for compatibility purposes.

StoneyJackson replied:
We'll burn that bridge when we get there. The important thing is to isolate OpenCV. That way we can easily figure out how to port it to v3.

StoneyJackson commented:
On June 11, 2014, Stoney Jackson was finally able to move the mouse pointer with his head using a webcam. He can now die a happy man.

kevin-brown commented on src/mousetrap/Makefile.am:
```diff
@@ -1,11 +1,17 @@
 
 install-exec-hook:
-	chmod a+rx $(DESTDIR)$(mousetrap_pythondir) -R
+	chmod a+r $(DESTDIR)$(mousetrap_pythondir) -R
```
For some reason, when running this on a fresh install (never ran old installer), this directory does not exist when this command is run, so an error is raised and the installer fails.

StoneyJackson replied:
Right, that makes sense now that I look at it again. I'm adding x back.

kevin-brown commented on src/mousetrap/main.py:
```diff
@@ -0,0 +1,41 @@
+    def __init__(self):
+        self.timeout_id = None
+        self.camera = Camera()
+        self.camera.set_dimensions(300, 200)
```
When testing earlier I was hitting issues with the image being this small.  After making the dimensions 600x400, the detection started working again and I didn't hit any extra issues.  I'm not sure why the image was chosen to be this size, so I was wondering if there was anything that came up?

StoneyJackson replied:
First, these dimensions are just suggestions. I think they are maximums. The actual size can be determined by the image read from the camera.

I did have some detection trouble, but I found that reducing the scale factor to 1.3 or even 1.1 helped. At this point, it works fairly well on my system, as long as I'm looking square, straight up-and-down, at the camera with good lighting.

If larger images help, we could capture and process larger images, and scale down images when we go to display them.

kevin-brown replied:
_(Continuing from IRC last night)_
We confirmed that my webcam was indeed returning a small image, 160x120 when we were requesting a 300x200 maximum image.  This was causing issues because, combined with my distance from the webcam, my nose was not able to be detected as it did not cover enough pixels (confirmed after moving closer).  After looking around a bit today, I found a StackOverflow post which cleared a lot of this up for me.

http://stackoverflow.com/a/20120262/359284

kevin-brown commented on src/mousetrap/gui.py:
```diff
@@ -0,0 +1,70 @@
+        # FIXME: Closing any window kills the application. Need a mechanism
+        # that allows windows to be openned and closed, and only kill the
+        # application when the last window is closed.
+        self._window.connect("delete-event", Gtk.main_quit)
```
`Gtk.main_quit` will kill all windows when you may be looking to just close the window the specific window.  Calling `self._window.destroy()` should do the job without closing the other windows.

StoneyJackson replied:
Good suggestion. I'm adding your comment to the issue.

kevin-brown commented:
It took us four days to go from moving the mouse to making a click. https://github.com/GNOME-MouseTrap/mousetrap/commit/6589e2d1a85ce4f2a0b272f3737b56521fa7f2a0

After extensive research, I have concluded that using PyGDK it will not be possible to send mouse events to the display.  I have decided to go back to using [`python-xlib`](http://python-xlib.sourceforge.net/), which is currently not Python 3 compatible.  While a Python 3 fork exists, [`python3-xlib`](https://github.com/LiuLang/python3-xlib), the package is not available in most repositories but it available through PyPI.

StoneyJackson committed:
Add support for nosetests.

StoneyJackson committed:
Remove nose dependency. Now only dependent on coverage (optionally).

StoneyJackson committed:
Fix options to coverage to include all and only mousetrap files.

StoneyJackson committed:
Rename test to match file.

StoneyJackson committed:
Remove cruft.

StoneyJackson committed:
Merge pull request #40 from GNOME-MouseTrap/detectcache

Add FeatureDetector caching.

StoneyJackson committed:
Initialize Loop using passed config and not app.config.

StoneyJackson committed:
Observable doesn't need a config.

StoneyJackson committed:
Report missing statements in coverage report.

StoneyJackson committed:
Add tests for mousetrap.core.Observable.

StoneyJackson committed:
Start test_config.py. Add test for _rmerge.

StoneyJackson committed:
Simplify mousetrap.config and add tests.

StoneyJackson committed:
Add more tests for mousetrap.config.

StoneyJackson committed:
Delete most of mt-*.

I'm was the only one using it, and I'm finding I can live without it.
So why have more code to maintain. Goodbye!

StoneyJackson committed:
PEP8/de-linting and adjust `make clean` to delete .coverage.

StoneyJackson committed:
Py3 compat: change from GObject.timeout_add to GLib.timeout_add.

`python -3` told me to.

StoneyJackson committed:
Port to Python 2.7+ (i.e., ready for Python 3).

StoneyJackson committed:
More port to Python 3.

kevin-brown committed:
More Python 3 compatibility fixes

kevin-brown committed:
`cv` no longer exists

StoneyJackson committed:
Convert comments to code.
